# Exercise 1

### 1.

In [1]:
import numpy as np

In [2]:
A = np.random.rand(6,6) # creates a random array of values between 0 and 1 in a 6x6 shape

In [3]:
A # prints array to make sure it works

array([[0.02619088, 0.36104403, 0.31129017, 0.42362089, 0.12894986,
        0.22245678],
       [0.46743626, 0.96315409, 0.23483198, 0.9035416 , 0.68133659,
        0.98825399],
       [0.64979274, 0.75853935, 0.61952911, 0.72217915, 0.30074408,
        0.62849498],
       [0.55869639, 0.5490333 , 0.42214116, 0.95030069, 0.23097357,
        0.06298601],
       [0.42719211, 0.05736204, 0.71913063, 0.09226325, 0.82600656,
        0.76253464],
       [0.1460198 , 0.80440129, 0.01655083, 0.37270676, 0.96737896,
        0.70949499]])

### 2.

In [4]:
print(A[2]) # prints third row of the array since the first row would be print(A[0])

[0.64979274 0.75853935 0.61952911 0.72217915 0.30074408 0.62849498]


### 3.

In [5]:
print(A[:,0::2]) # prints all of the values in the row portion with [:] 
                 # and then starts with the first column and goes by every 
                 # 2 to get all of the odd columns {,0::2}

[[0.02619088 0.31129017 0.12894986]
 [0.46743626 0.23483198 0.68133659]
 [0.64979274 0.61952911 0.30074408]
 [0.55869639 0.42214116 0.23097357]
 [0.42719211 0.71913063 0.82600656]
 [0.1460198  0.01655083 0.96737896]]


### 4.

In [6]:
number_dictionary = {'even': 0, 'odd': 0} # creates a dictionary and initially asigns integer value 0 to even and odd

In [7]:
for i in range(A.shape[0]): # i is now rows
    for j in range(A.shape[1]): # j is now columns
        if (i + j) % 2 == 0:
            number_dictionary['even'] += A[i,j] # if the condition is met then it will add the values in the even coordinates
        else:
            number_dictionary['odd'] += A[i,j] # if the condition is not met then it will add the values in the odd coordinates

In [8]:
print(number_dictionary)

{'even': 10.312698867833316, 'odd': 7.753860646670889}


### 5.

In [47]:
class matrixMaker: # creates a class titled matrixMaker
    def __init__(self,m,n):
        self.m = m
        self.n = n
        self.A = np.random.rand(m,n) # random array with rows = m and columns = n
    
    def specialSum(self):
        return np.sum(self.A[self.A > 0.5]) # sum of the values in the array that are greater than 0.5

In [48]:
test = matrixMaker(5,5) # this test creates a 5x5 matrix 

In [49]:
print(test.specialSum()) # this prints the specialSum of the test matrix

11.813530206936576


# Exercise 2

### 1.

In [52]:
import nltk

In [53]:
plurals = ['caresses', 'flies', 'dies', 'mules', 'denied',
...            'died', 'agreed', 'owned', 'humbled', 'sized',
...            'meeting', 'stating', 'siezing', 'itemization',
...            'sensational', 'traditional', 'reference', 'colonizer',
...            'plotted']

### 2.

In [54]:
from nltk.stem import PorterStemmer , WordNetLemmatizer

In [57]:
stem = PorterStemmer() # calls in PorterStemmer to then use below

In [58]:
stems = [stem.stem(word) for word in plurals] # all of the stems in plurals assigned to stems

In [59]:
print(stems)

['caress', 'fli', 'die', 'mule', 'deni', 'die', 'agre', 'own', 'humbl', 'size', 'meet', 'state', 'siez', 'item', 'sensat', 'tradit', 'refer', 'colon', 'plot']


### 3.

In [60]:
lemmatize = WordNetLemmatizer()

In [62]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to C:\Users\Danny
[nltk_data]     Tapp\AppData\Roaming\nltk_data...


True

In [64]:
nltk.download('omw-1.4') # realized I had to download wordnet and omw-1.4 when I tried to use WordNetLemmatizer

[nltk_data] Downloading package omw-1.4 to C:\Users\Danny
[nltk_data]     Tapp\AppData\Roaming\nltk_data...


True

In [65]:
lemmatized = [lemmatize.lemmatize(word) for word in plurals]

In [66]:
print(lemmatized)

['caress', 'fly', 'dy', 'mule', 'denied', 'died', 'agreed', 'owned', 'humbled', 'sized', 'meeting', 'stating', 'siezing', 'itemization', 'sensational', 'traditional', 'reference', 'colonizer', 'plotted']


### 4.

In [67]:
from nltk.tokenize import word_tokenize

In [68]:
text = "Wow, we can finally stop using the split function!"

In [71]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to C:\Users\Danny
[nltk_data]     Tapp\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.


True

In [75]:
tokens = word_tokenize(text) # tokenized the sample text and assigned it to tokens

In [76]:
print(tokens)

['Wow', ',', 'we', 'can', 'finally', 'stop', 'using', 'the', 'split', 'function', '!']


# Exercise 3

### 1.

In [114]:
with open("nemo.txt", "r") as nemo_script: # opened up the text file (which was in the same folder as this notebook)
    data = nemo_script.readlines()
    
total_words = 0 # assigned it to 0 to begin with
    
for line in data:
    words = line.split() # this will get each word
    total_words += len(words) # this will add up the words 

### 2.

In [115]:
print(total_words) 

12263


### 3.

In [133]:
from collections import defaultdict # Needed a default dictionary instead of dict because it has default values and I won't get KeyError

each_chr_word_count = defaultdict(int) # This will create a default value of 0 when we start counting a new character's word count

with open("nemo.txt","r") as nemo_script:
    data = nemo_script.readlines() # Opened up the text file
    current_character = None # = None means that there's an absence of a value instead of 0 which is an integer,
                             # value will be assigned later
    for line in data:
        line = line.strip() # removes leading and trailling whitespace characters from a string
        if line.isupper(): # all character names were in full uppercase, so this checks for character names
            current_character = line
        elif current_character:
            words = line.split() # separates the words in the lines
            each_chr_word_count[current_character] += len(words) # adds the total words to the dictionary for the character
            
sorted_word_counts = sorted(each_chr_word_count.items(),key=lambda item: item[1], reverse=True) # ordered the word count in descending order

for character, count in sorted_word_counts:
    print(f"{character}: {count} words") # f string to make it easier to print character names and corresponding word counts
    
    
    

MARLIN: 3357 words
DORY: 1816 words
GILL: 782 words
NEMO: 661 words
NIGEL: 430 words
DENTIST: 324 words
MR. RAY: 321 words
PEACH: 283 words
BRUCE: 279 words
CRUSH: 272 words
BLOAT: 251 words
GURGLE: 194 words
MOONFISH LEADER: 188 words
CHUM: 129 words
CORAL: 128 words
DEB: 122 words
SQUIRT: 92 words
ANCHOR: 68 words
BOB: 57 words
DARLA: 52 words
TAD: 51 words
PELICAN: 48 words
CRAB 2: 48 words
PEARL: 47 words
BUBBLES: 47 words
BLOAT/BUBBLES/GURGLE: 47 words
GULLS: 47 words
BILL: 42 words
SHELDON: 33 words
LOBSTER: 31 words
BRUCE/ANCHOR/CHUM: 30 words
FISH: 27 words
JACQUES: 26 words
SWORDFISH: 24 words
CRAB 1: 22 words
TURTLE KID: 18 words
FEMALE BIRD: 18 words
ALL: 18 words
CRAB 1/CRAB 2: 18 words
SMALL FISH: 17 words
MALE BIRD 1: 17 words
MR. JOHANSSEN: 16 words
DOLPHIN: 16 words
CRAB: 16 words
BARBARA: 15 words
PATIENT: 15 words
MALE BIRD 2: 12 words
MARLIN/DORY: 10 words
CRAB KID: 9 words
TED: 9 words
BOB/TED: 9 words
KIDS: 8 words
FISH KID: 8 words
TURTLE KIDS: 8 words
TURTLE KID 

#### Marlin spoke the most with a total of 3357 words